src/wallet/spend.cpp

```diff
@@ -153,6 +153,31 @@ BResult<CoinsResult> FetchSelectedInputs(const CWallet& wallet, const CCoinContr
     return result;
 }
 
+BResult<CoinsFund> FetchCoins(const CWallet& wallet, const CCoinControl& coin_control,
+                              const CoinSelectionParams& coin_selection_params) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet)
+{
+    CoinsFund result;
+
+    // Fetch manually select coins
+    if (coin_control.HasSelected()) {
+        result.pre_selected_inputs = FetchSelectedInputs(wallet, coin_control, coin_selection_params);
+        if (!result.pre_selected_inputs) return result.pre_selected_inputs.GetError();
+
+        // If no other inputs are allowed, return early
+        if (!coin_control.m_allow_other_inputs) return result;
```

Comment:
M
In 84b4ffe85ba0267c49080e9a096e45d6af63e720 "wallet: remove fetch pre-selected-inputs responsibility from SelectCoins"

This check could be moved to the top of `AvailableCoins` and it would achieve the same effect. Doing so would also not change `AvailableCoins`'s semantics.

Reply:
I added it here (inside the `coin_control.HasSelected()` block) to not change the current behavior and minimize the amount of changes.

In master, even if `m_allow_other_inputs` is false, we are automatically falling back to select coins from the wallet if there are no pre-set inputs inside coin control.

This basically is because the default value for `m_allow_other_inputs` is false, and some RPC commands that don't allow inputs manual selection such as `sendtoaddress` and `sendmany` are not setting it to true. So.. they are always falling back to select coins from the wallet.

So.. if I move this check below, will need to set `m_allow_other_inputs=true` in every caller that does not add any input manually (or.. change the default value to be true inside coin control).

but.. let me see if those two RPC commands are the only affected ones. Maybe it's a straightforward change.
I do agree that the current master behavior of automatically falling back to select coins from the wallet even when `m_allow_other_inputs=false` is messy and not an expected behavior.